crypto/asn1/asn1_par.c

```diff
@@ -101,3 +118,527 @@ const char *ASN1_tag2str(int tag) {
   }
   return tag2str[tag];
 }
+
+int ASN1_parse(BIO *bp, const unsigned char *pp, long len, int indent) {
+  GUARD_PTR(bp);
+  GUARD_PTR(pp);
+  return asn1_parse2(bp, &pp, len, 0, 0, indent, 0);
+}
+
+// Constants
+#define ASN1_PARSE_MAXDEPTH 128
+#define ASN1_DUMP_INDENT 6  // Because we know BIO_dump_indent()
+
+// Copy of helper functions from original (these remain unchanged)
+static int asn1_print_info(BIO *bp, int tag, int xclass, int constructed,
+                           int indent) {
+  static const char fmt[] = "%-18s";
+  char str[128];
+  const char *p;
+
+  if (constructed & V_ASN1_CONSTRUCTED) {
+    p = "cons: ";
+  } else {
+    p = "prim: ";
+  }
+  if (BIO_write(bp, p, 6) < 6) {
+    goto err;
+  }
+  BIO_indent(bp, indent, 128);
+
+  p = str;
+  if ((xclass & V_ASN1_PRIVATE) == V_ASN1_PRIVATE) {
+    BIO_snprintf(str, sizeof(str), "priv [ %d ] ", tag);
+  } else if ((xclass & V_ASN1_CONTEXT_SPECIFIC) == V_ASN1_CONTEXT_SPECIFIC) {
+    BIO_snprintf(str, sizeof(str), "cont [ %d ]", tag);
+  } else if ((xclass & V_ASN1_APPLICATION) == V_ASN1_APPLICATION) {
+    BIO_snprintf(str, sizeof(str), "appl [ %d ]", tag);
+  } else if (tag > 30) {
+    BIO_snprintf(str, sizeof(str), "<ASN1 %d>", tag);
+  } else {
+    p = ASN1_tag2str(tag);
+  }
+
+  if (BIO_printf(bp, fmt, p) <= 0) {
+    goto err;
+  }
+  return 1;
+err:
+  return 0;
+}
+
+static int BIO_dump_indent(BIO *bp, const char *s, int len, int indent) {
+  // Use BIO_hexdump as a replacement for BIO_dump_indent
+  return BIO_hexdump(bp, (const uint8_t *)s, len, indent);
+}
+
+// Helper function to parse constructed ASN.1 types (SEQUENCE, SET, etc.)
+static int asn1_parse_constructed_type(BIO *bp, const uint8_t **current_pos,
+                                       const uint8_t *total_end,
+                                       const uint8_t *original_start,
+                                       long *object_len, int parse_flags,
+                                       int offset, int depth, int indent,
+                                       int dump) {
+  GUARD_PTR(bp);
+  GUARD_PTR(current_pos);
+  GUARD_PTR(total_end);
+  GUARD_PTR(original_start);
+  GUARD_PTR(object_len);
+
+  const uint8_t *start_pos = *current_pos;
+
+  if (BIO_write(bp, "\n", 1) <= 0) {
+    return 0;
+  }
+
+  if ((parse_flags == (V_ASN1_CONSTRUCTED | 1)) && (*object_len == 0)) {
+    // Indefinite length constructed object
+    for (;;) {
+      const int parse_result = asn1_parse2(
+          bp, current_pos, (long)(total_end - *current_pos),
+          offset + (*current_pos - original_start), depth + 1, indent, dump);
```

Comment:
Although unlikely, it's possible for `offset + (*current_pos - original_start)` to overflow here. We could check on the line above, but I think it might be better to add a few "sanity checks" at the beginning of `asn1_parse2` to verify that values like `length`, `offset` and `depth` are non-negative.